### PR TITLE
Introduce PhysicalServerProvision StateMachine

### PR DESCRIPTION
**Pull request by miha-plesko**
With this commit we implement a new external state machine
    
```
PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision
```
    
which in essence invokes (also new) internal state machine and then polls for    its status. The external state machine uses AE messages to allow for different     customization for redfish-managed servers than for lenovo-manager ones.

Depends on 
https://github.com/ManageIQ/manageiq/pull/18573
https://github.com/ManageIQ/manageiq-automation_engine/pull/306

@miq-bot add_label enhancement

### CODE REVIEW

File: content/automate/ManageIQ/Physical_Infrastructure/PhysicalServer/Lifecycle.class/provisioning.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+version: 1.0
+object:
+  attributes:
+    display_name: 
+    name: Provisioning
+    inherits: 
+    description: 
+  fields:
+  - Relationship4:
+      value: "/ManageIQ/System/Request/InspectMe"
```
Comment (lfu) [inline content/automate/ManageIQ/Physical_Infrastructure/PhysicalServer/Lifecycle.class/provisioning.yaml]: This is useful for debugging. But should not be in production code.

File: content/automate/ManageIQ/Physical_Infrastructure/PhysicalServer/Lifecycle.class/provisioning.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  - Relationship5:
+      value: "/Physical_Infrastructure/PhysicalServer/Provisioning/Profile/${/#user.normalized_ldap_group}#get_state_machine"
```
Comment (lfu) [inline content/automate/ManageIQ/Physical_Infrastructure/PhysicalServer/Lifecycle.class/provisioning.yaml]: All ```Physical_Infrastructure``` should be changed to ```PhysicalInfrastructure```.

File: content/automate/ManageIQ/Physical_Infrastructure/PhysicalServer/Provisioning/StateMachines/Methods.class/__methods__/common_provision.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+$evm.log(:info, "server = #{server.id}")
+$evm.log(:info, "server.vendor = #{server.vendor}")
+$evm.log(:info, "server.hostname = #{server.hostname}")
+$evm.log(:info, "server.kind = #{server.kind}")
+
+$evm.log(:info, "options = #{options}")
+$evm.log(:info, "pxe_server_id = #{options[:pxe_server_id]}")
+$evm.log(:info, "src_pxe_image_id = #{options[:src_pxe_image_id]}")
+$evm.log(:info, "src_configuration_profile_id = #{options[:src_configuration_profile_id]}")
```
Comment (lfu) [inline content/automate/ManageIQ/Physical_Infrastructure/PhysicalServer/Provisioning/StateMachines/Methods.class/__methods__/common_provision.rb]: For the newer methods we are switching to a new format which helps us to test the methods easily.

For example you can look at https://github.com/ManageIQ/manageiq-content/blob/2b6360cc885c7f099838c84c121b8fe750ffb8ba/content/automate/ManageIQ/Cloud/VM/Provisioning/Placement.class/__methods__/best_fit_amazon.rb

File: content/automate/ManageIQ/Physical_Infrastructure/PhysicalServer/Provisioning/StateMachines/Methods.class/__methods__/redfish_provision.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  server.provision(
+    options[:pxe_server_id],
+    options[:src_pxe_image_id],
+    options[:src_configuration_profile_id]
+  )  
+  exit MIQ_OK
+rescue => err  
+  $evm.log(:error, "[#{err}]\n#{err.backtrace.join("\n")}")
+  $evm.root['ae_result'] = 'error'
+  exit MIQ_STOP
```
Comment (lfu) [inline content/automate/ManageIQ/Physical_Infrastructure/PhysicalServer/Provisioning/StateMachines/Methods.class/__methods__/redfish_provision.rb]: Automate does the error handling for these methods.

File: content/automate/ManageIQ/Physical_Infrastructure/PhysicalServer/Provisioning/StateMachines/PhysicalServerProvision.class/__methods__/update_provision_status.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+updated_message += "Status [#{status}] "
+updated_message += "Message [#{prov.message}] "
+updated_message += "Current Retry Number [#{$evm.root['ae_state_retries']}]" if $evm.root['ae_result'] == 'retry'
+prov.miq_request.user_message = updated_message
+
+if $evm.root['ae_result'] == "error"
+  $evm.create_notification(:level => "error", :subject => prov.miq_request, \
+                           :message => "PhysicalServer Provision Error: #{updated_message}")
+  $evm.log(:error, "PhysicalServer Provision Error: #{updated_message}")
+end
```
Comment (lfu) [inline content/automate/ManageIQ/Physical_Infrastructure/PhysicalServer/Provisioning/StateMachines/PhysicalServerProvision.class/__methods__/update_provision_status.rb]: Please convert to new format.

File: content/automate/ManageIQ/Physical_Infrastructure/PhysicalServer/Provisioning/StateMachines/Methods.class/__methods__/default_provision.rb [hunk 6]
```diff
@@ -0,0 +1,4 @@
+#
+# Description: <Method description here>
+#
+$evm.log(:info, "BUREK PROVISIONING DEFAULT PHYSICAL SERVER")
```
Comment (lfu) [inline content/automate/ManageIQ/Physical_Infrastructure/PhysicalServer/Provisioning/StateMachines/Methods.class/__methods__/default_provision.rb]: Same here for new format.

File: content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/PhysicalServerProvision.class/__methods__/update_provision_status.rb [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+      module PhysicalServer
+        module Provisioning
+          module StateMachines
+            module PhysicalServerProvision
+              class UpdateProvisionStatus
+                def initialize(handle = $evm)
+                  @handle = handle
+                end
+
+                def main
```
Comment (lfu) [inline content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/PhysicalServerProvision.class/__methods__/update_provision_status.rb]: Please break this into small methods so each one may be around 6~7 lines.

File: content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/PhysicalServerProvision.class/__methods__/update_provision_status.rb [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+            end
+          end
+        end
+      end
+    end
+  end
+end
+
+ManageIQ::Automate::PhysicalInfrastructure::PhysicalServer::Provisioning::StateMachines::PhysicalServerProvision::UpdateProvisionStatus.new.main
+
```
Comment (lfu) [inline content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/PhysicalServerProvision.class/__methods__/update_provision_status.rb]: Please remove the empty lines.

**Comment by miha-plesko** [comment]
Thanks for reviewing @lfu , I've pushed an updated version with your suggestions applied plus unit tests for all the new methods.

BTW I'm wondering whether we need to change anything in order to support scenario where single Request would be realized with more than one Task? If I create request with, say, three source ids

```
src_id = [1, 2, 3]
```

then the three servers will get provisioned, but there seems to be "race condition" between them as they are all updating same request's status. Is there perhaps any example on how to handle such scenario?

**Comment by miha-plesko** [comment]
Marking as WIP again, because we need to merge core changes first or else Travis goes crazy

File: content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Methods.class/__methods__/check_provision.rb [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+                  result = task.statemachine_task_status
+                  server = task.source
+                  server_str = "PhysicalServer id=#{server.id} ems_ref=#{server.ems_ref}"
+
+                  @handle.log('info', "ProvisionCheck (#{server_str}) returned <#{result}> for state <#{task.state}> and status <#{task['status']}>")
+
+                  case result
+                  when 'error'
+                    @handle.root['ae_result'] = 'error'
+                    @handle.root['ae_reason'] = task.message.to_s.sub('Error: ', '')
```
Comment (lfu) [inline content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Methods.class/__methods__/check_provision.rb]: Why do we need ```to_s```? ```task.message``` returns a string.

File: content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/__methods__/update_provision_status.rb [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+                    unless task
+                      @handle.log(:error, 'miq_provision_task object not provided')
+                      exit(MIQ_STOP)
+                    end
+                    unless task.source
+                      @handle.log(:error, 'miq_provision_task.source object not provided (PhysicalServer expected)')
+                      exit(MIQ_STOP)
+                    end
+                    task
+                  end
```
Comment (lfu) [inline content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/__methods__/update_provision_status.rb]:
```
@task ||= @handle.root['miq_provision_task'].tap do |task|
  unless task
    @handle.log(:error, 'miq_provision_task object not provided')
    exit(MIQ_STOP)
  end
  unless task.source
    @handle.log(:error, 'miq_provision_task.source object not provided (PhysicalServer expected)')
    exit(MIQ_STOP)
  end
end
```

File: spec/content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Methods.class/__methods__/provision_spec.rb [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+  let(:admin)       { FactoryBot.create(:user_admin) }
+  let(:server)      { FactoryBot.create(:physical_server) }
+  let(:svc_server)  { MiqAeMethodService::MiqAeServicePhysicalServer.find_by(:id => server.id) }
+  let(:request)     { FactoryBot.create(:physical_server_provision_request, :requester => admin) }
+  let(:task)        { FactoryBot.create(:physical_server_provision_task, :miq_request => request, :source => server) }
+  let(:svc_task)    { MiqAeMethodService::MiqAeServiceMiqProvisionTask.find_by(:id => task.id) }
+  let(:root_object) { Spec::Support::MiqAeMockObject.new(:miq_provision_task => svc_task) }
+  let(:ae_service)  { Spec::Support::MiqAeMockService.new(root_object) }
+
+  it 'execute is called on task' do
```
Comment (lfu) [inline spec/content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Methods.class/__methods__/provision_spec.rb]:
```
it 'calls execute on task' do
```

File: spec/content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/__methods__/update_provision_status_spec.rb [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+  let(:root_object) { Spec::Support::MiqAeMockObject.new(:miq_provision_task => svc_task) }
+  let(:ae_service)  { Spec::Support::MiqAeMockService.new(root_object) }
+
+  subject { described_class.new(ae_service) }
+
+  describe '#status_param' do
+    before { ae_service.inputs['status'] = 'the status' }
+    it do
+      expect(subject.status_param).to eq('the status')
+    end
```
Comment (lfu) [inline spec/content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/__methods__/update_provision_status_spec.rb]:
```
it { expect(subject.status_param).to eq('the status') }
```

File: spec/content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/__methods__/update_provision_status_spec.rb [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+  end
+
+  describe '#main' do
+    before do
+      ae_service.inputs['status'] = 'the status'
+      root_object['miq_server'] = double('MIQ_SERVER', :name => 'miq server name')
+      root_object['ae_state'] = 'at state'
+    end
+    context 'updates task and request message' do
+      it do
```
Comment (lfu) [inline spec/content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/__methods__/update_provision_status_spec.rb]: Please add a description for ```it```.

File: spec/content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/__methods__/update_provision_status_spec.rb [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+        expect(request_msg).to include('[miq server name]')
+        expect(request_msg).to include("PhysicalServer [#{server.id}|#{server.ems_ref}]")
+        expect(request_msg).to include('Step [at state]')
+        expect(request_msg).to include('Message [the status]')
+      end
+    end
+
+    context 'fires notification on error' do
+      before { root_object['ae_result'] = 'error' }
+      it do
```
Comment (lfu) [inline spec/content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/__methods__/update_provision_status_spec.rb]: Same here.

File: content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Methods.class/__methods__/check_provision.rb (hunk 9 above)
Comment (miha-plesko) [reply content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Methods.class/__methods__/check_provision.rb]: Good point

File: content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/__methods__/update_provision_status.rb (hunk 10 above)
Comment (miha-plesko) [reply content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/__methods__/update_provision_status.rb]: You know, it felt like there is a better way, thanks 👍

File: spec/content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Methods.class/__methods__/provision_spec.rb (hunk 11 above)
Comment (miha-plesko) [reply spec/content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Methods.class/__methods__/provision_spec.rb]: Done

File: spec/content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/__methods__/update_provision_status_spec.rb (hunk 13 above)
Comment (miha-plesko) [reply spec/content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/__methods__/update_provision_status_spec.rb]: Done

File: spec/content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/__methods__/update_provision_status_spec.rb (hunk 14 above)
Comment (miha-plesko) [reply spec/content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/__methods__/update_provision_status_spec.rb]: Done

**Comment by miha-plesko** [comment]
Sorry @lfu had to change it a little bit: I've replaced all occurances of

```
$evm.root['miq_provision_task']
```

with

```
$evm.root['physical_server_provision_task']
```

because we've modified automation engine a bit, see here https://github.com/ManageIQ/manageiq/pull/18573/files#r277710290

File: content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/default.yaml [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+version: 1.0
+object:
+  attributes:
+    display_name: 
+    name: default
+    inherits: 
+    description: 
+  fields:
+  - PreProvision:
+      value: "/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Methods/PreProvision#${/#miq_provision_task.source.ems_type}"
```
Comment (lfu) [inline content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/default.yaml]: Is ```miq_provision_task``` used in this file by purpose?

Comment (miha-plesko) [reply content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/default.yaml]: Thanks

**Comment by lfu** [comment]
LGTM. Please remove WIP from the title when PR is ready to merge.

**Comment by miha-plesko** [comment]
@lfu slightly updated again following @gmcculloug 's advice here https://github.com/ManageIQ/manageiq/pull/18573#discussion_r279118400. This time I replaced

```
source.ems_type -> source.emstype
```

**Comment by lfu** [comment]
LGTM 👍 Remember to remove WIP from the title when PR is ready to merge.

**Pull request merged**